pkg/kritis/review/review.go

```diff
@@ -52,25 +85,144 @@ func (r Reviewer) Review(images []string, isps []v1beta1.ImageSecurityPolicy, po
 			}
 			if len(violations) != 0 {
-				errMsg := fmt.Sprintf("found violations in %s", image)
-				// Check if one of the violations is that the image is not fully qualified
-				for _, v := range violations {
-					if v.Violation == securitypolicy.UnqualifiedImageViolation {
-						errMsg = fmt.Sprintf("%s is not a fully qualified image", image)
-					}
+				return r.handleViolations(image, pod, violations)
+			} else if r.isWebhook {
+				if err := r.addAttestations(image, attestations, isp.Namespace); err != nil {
+					glog.Infof("error adding attestations %s", err)
 				}
-				if err := r.vs.HandleViolation(image, pod, violations); err != nil {
-					return fmt.Errorf("%s. error handling violation %v", errMsg, err)
-				}
-				return fmt.Errorf(errMsg)
 			}
 		}
 	}
 	return nil
 }
+
+func (r Reviewer) fetchAndVerifyAttestations(image string, ns string, pod *v1.Pod) (bool, []metadata.PGPAttestation) {
+	attestations, err := r.client.GetAttestations(image)
+	if err != nil {
+		glog.Infof("Error while fetching attestations %s", err)
+		return false, attestations
+	}
+	isAttested := r.hasValidImageAttestations(image, attestations, ns)
+	if err := r.vs.HandleAttestation(image, pod, isAttested); err != nil {
+		glog.Infof("error handling attestations %v", err)
+	}
+	return isAttested, attestations
+}
+
+// hasValidImageAttestations return true if any one image attestation is verified.
+func (r Reviewer) hasValidImageAttestations(image string, attestations []metadata.PGPAttestation, ns string) bool {
+	if len(attestations) == 0 {
+		glog.Infof(`No attestations found for this image.
+This normally happens when you deploy a pod before kritis or no attestation authority is deployed.
+Please see instructions `)
+	}
+	host, err := container.NewAtomicContainerSig(image, map[string]string{})
+	if err != nil {
+		glog.Info(err)
+		return false
+	}
+	for _, a := range attestations {
+		// Get Secret from key id.
+		secret, err := r.config.secret(ns, a.KeyId)
+		if err != nil {
+			glog.Infof("Could not find secret %s in namespace %s for attestation verification", a.KeyId, ns)
+			continue
+		}
+		if err = host.VerifyAttestationSignature(secret.PublicKey, a.Signature); err != nil {
+			glog.Infof("Could not find verify attestation for attestation authority %s", a.KeyId)
+		} else {
+			return true
+		}
+	}
+	return false
+}
+
+func (r Reviewer) handleViolations(image string, pod *v1.Pod, violations []securitypolicy.SecurityPolicyViolation) error {
+	errMsg := fmt.Sprintf("found violations in %s", image)
+	// Check if one of the violations is that the image is not fully qualified
+	for _, v := range violations {
+		if v.Violation == securitypolicy.UnqualifiedImageViolation {
+			errMsg = fmt.Sprintf("%s is not a fully qualified image", image)
+		}
+	}
+	if err := r.vs.HandleViolation(image, pod, violations); err != nil {
+		return fmt.Errorf("%s. error handling violation %v", errMsg, err)
```

Comment:
Remove errMsg - it doesn't seem relevant to this message. Y

Reply:
How about:

return fmt.Errorf("error handling violation %+s: %v", violations, err)

I don't think errMsg is relevant to the error, and it seems like an exceptional error where having the struct displayed may be of use to the debugger.